### PR TITLE
Add support for crio drop-in configuration files

### PR DESCRIPTION
**Pull request by saschagrunert**
We now support partial drop-in configuration files which will be assembled together on CRI-O startup.

A default (or user specified via the CLI) directory will be recursively iterated, whereas each file gets applied to the configuration in their processing order. This means that a configuration file named `/etc/crio/crio.conf.d/00-default` has a lower priority than a file `/etc/crio/crio.conf.d/01-my-config`. The configuration provided via `--config,-c` has always a higher priority than those inside the `--config-dir,-d`. The command line parameters have the highest configuration priority at all.

/cc @ereslibre @flavio @Martin-Weiss

### CODE REVIEW

**Comment by Martin-Weiss** [comment]
> ult (or user specified via the CL

How do you plan to migrate from customized crio.conf to drop-in configuration files?
Do you plan to change the spec file for crio.conf to overwrite it instead of the creation of "rpmnew"?

**Comment by saschagrunert** [comment]
> How do you plan to migrate from customized crio.conf to drop-in configuration files?
> Do you plan to change the spec file for crio.conf to overwrite it instead of the creation of "rpmnew"?

The plan would be that we do not use `crio.conf` any more and therefore provide a sane default in `/etc/crio/crio.conf.d/00-default`. Then users can customize their setup via `/etc/crio/crio.conf.d/01-change-something` et cetera.

**Comment by rhatdan** [comment]
I wonder how we can merge these with the work we are doing on containers.conf for podman/buildah.

**Comment by Martin-Weiss** [comment]
> 
> 
> > How do you plan to migrate from customized crio.conf to drop-in configuration files?
> > Do you plan to change the spec file for crio.conf to overwrite it instead of the creation of "rpmnew"?
> 
> The plan would be that we do not use `crio.conf` any more and therefore provide a sane default in `/etc/crio/crio.conf.d/00-default`, which is owned by the package. Then users can customize their setup via `/etc/crio/crio.conf.d/01-change-something` et cetera.

But in this case an upgrade will "remove" all customer made adjustments.

The current problem is that a customer had to configure global_auth - so with this change and ignoring the settings in crio.conf - the authentication to the registry will be gone after the upgrade and the next pull will fail.

So what ever change is done - IMO we also have to take care about what was there before and we have to migrate somehow.

**Comment by vrothberg** [comment]
> This means that a configuration file named /etc/crio/crio.conf.d/00-default has a lower priority than a file /etc/crio/crio.conf.d/01-my-config.

Does it mean a reverse alphanumerical order?

File: test/helpers.bash [hunk 1]
```diff
@@ -197,6 +197,8 @@ function setup_test() {
 	fi
 	CRIO_SOCKET="$TESTDIR/crio.sock"
 	CRIO_CONFIG="$TESTDIR/crio.conf"
+	CRIO_CONFIG_DIR="$TESTDIR/crio.conf.d"
+    mkdir "$CRIO_CONFIG_DIR"
```
Comment (haircommander) [inline test/helpers.bash]: nit: indentation

**Comment by haircommander** [comment]
> > > How do you plan to migrate from customized crio.conf to drop-in configuration files?
> > > Do you plan to change the spec file for crio.conf to overwrite it instead of the creation of "rpmnew"?
> > 
> > 
> > The plan would be that we do not use `crio.conf` any more and therefore provide a sane default in `/etc/crio/crio.conf.d/00-default`, which is owned by the package. Then users can customize their setup via `/etc/crio/crio.conf.d/01-change-something` et cetera.
> 
> But in this case an upgrade will "remove" all customer made adjustments.
> 
> The current problem is that a customer had to configure global_auth - so with this change and ignoring the settings in crio.conf - the authentication to the registry will be gone after the upgrade and the next pull will fail.
> 
> So what ever change is done - IMO we also have to take care about what was there before and we have to migrate somehow.

theoretically, we know what the defaults are and what the user changed values are. Upon rpm upgrade, we could have a command in the crio binary that translates between crio.conf to default-00 + changes-01

**Comment by saschagrunert** [comment]
> > > How do you plan to migrate from customized crio.conf to drop-in configuration files?
> > > Do you plan to change the spec file for crio.conf to overwrite it instead of the creation of "rpmnew"?
> > 
> > 
> > The plan would be that we do not use `crio.conf` any more and therefore provide a sane default in `/etc/crio/crio.conf.d/00-default`, which is owned by the package. Then users can customize their setup via `/etc/crio/crio.conf.d/01-change-something` et cetera.
> 
> But in this case an upgrade will "remove" all customer made adjustments.
>
> The current problem is that a customer had to configure global_auth - so with this change and ignoring the settings in crio.conf - the authentication to the registry will be gone after the upgrade and the next pull will fail.
> 
> So what ever change is done - IMO we also have to take care about what was there before and we have to migrate somehow.

Yes. We already discussed upgrade migration features in https://github.com/cri-o/cri-o/issues/2579 but I'm open to re-visit that discussion.

> > This means that a configuration file named /etc/crio/crio.conf.d/00-default has a lower priority than a file /etc/crio/crio.conf.d/01-my-config.
> 
> Does it mean a reverse alphanumerical order?

Yes, we apply the configuration in filename order, so the priority is reversed from low to high. Does this make sense?

File: test/helpers.bash (hunk 1 above)
Comment (saschagrunert) [reply test/helpers.bash]: Fixed :+1:

**Comment by mrunalp** [comment]
I think we should align with how systemd does some of this: https://www.freedesktop.org/software/systemd/man/systemd-system.conf.html.

One thing where we are differing in this implementation vs. systemd is in priority of crio.conf. I think t the files in conf.d should still allow overriding crio.conf. 

Thanks for putting this together. This has been on my list for so long :)

**Comment by saschagrunert** [comment]
> One thing where we are differing in this implementation vs. systemd is in priority of crio.conf. I think the files in conf.d should still allow overriding crio.conf.

Alright :+1:, I changed the configuration priority to this order and adapted the documentation:
1. CLI arguments
1. `crio.conf.d/*`
1. `crio.conf`

File: docs/crio.8.md [hunk 2]
```diff
@@ -125,6 +126,17 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
 
 **--config, -c**="": Path to configuration file (default: /etc/crio/crio.conf)
 
+**--config-dir, -d**="": Path to the configuration drop-in directory.
+    This directory will be recursively iterated and each file gets applied
+    to the configuration in their processing order. This means that a
+    configuration file named '00-default' has a lower priority than a file
+    '01-my-overwrite'.
+    The global config file, either provided via '--config,-c' or per default in
+    /etc/crio/crio.conf, has always a lower priority than those inside the
```
Comment (TomSweeneyRedHat) [inline docs/crio.8.md]: "has always" to "always has"

File: docs/crio.8.md [hunk 3]
```diff
@@ -125,6 +126,17 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
 
 **--config, -c**="": Path to configuration file (default: /etc/crio/crio.conf)
 
+**--config-dir, -d**="": Path to the configuration drop-in directory.
+    This directory will be recursively iterated and each file gets applied
+    to the configuration in their processing order. This means that a
+    configuration file named '00-default' has a lower priority than a file
+    '01-my-overwrite'.
```
[the remark this answers is not in the document]
Comment (saschagrunert) [reply docs/crio.8.md]: Fixed as suggested :+1:

File: docs/crio.8.md (hunk 2 above)
Comment (saschagrunert) [reply docs/crio.8.md]: Fixed as well :+1:

File: docs/crio.8.md [hunk 4]
```diff
@@ -125,6 +126,17 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
...
+    The global config file, either provided via '--config,-c' or per default in
```
Comment (TomSweeneyRedHat) [inline docs/crio.8.md]: I'd drop 'either' in this line, it's a bit redundant with the 'or'

File: docs/crio.8.md [hunk 5]
```diff
@@ -125,6 +126,17 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
...
+    /etc/crio/crio.conf, has always a lower priority than those inside the
+    '--config-dir,-d'.
```
Comment (TomSweeneyRedHat) [inline docs/crio.8.md]: suggest:  "than those inside the '--config-dir,-d'." to "than the files in the directory specified by '--config-dir,-d'."

File: docs/crio.8.md [hunk 6]
```diff
@@ -125,6 +126,17 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
...
+    Beside that, the provided command line parameters still have the highest
+    configuration priority. (default: /etc/crio/crio.conf.d)
```
Comment (TomSweeneyRedHat) [inline docs/crio.8.md]: I'm not sure what this last bit is trying to explain.

File: completions/fish/crio.fish [hunk 7]
```diff
@@ -16,6 +16,16 @@ complete -c crio -n '__fish_crio_no_subcommand' -f -l cgroup-manager -r -d 'cgro
...
+complete -c crio -n '__fish_crio_no_subcommand' -l config-dir -s d -r -d 'Path to the configuration drop-in directory.
+    This directory will be recursively iterated and each file gets applied
+    to the configuration in their processing order. This means that a
+    configuration file named \'00-default\' has a lower priority than a file
+    \'01-my-overwrite\'.
+    The global config file, either provided via \'--config,-c\' or per default in
+    /etc/crio/crio.conf, has always a lower priority than those inside the
+    \'--config-dir,-d\'.
+    Beside that, the provided command line parameters still have the highest
+    configuration priority.'
```
Comment (TomSweeneyRedHat) [inline completions/fish/crio.fish]: see below comments.

File: internal/pkg/criocli/criocli.go [hunk 8]
```diff
@@ -289,6 +294,23 @@ func getCrioFlags(defConf *libconfig.Config, systemContext *types.SystemContext)
...
+			Usage: fmt.Sprintf("Path to the configuration drop-in directory."+`
+    This directory will be recursively iterated and each file gets applied
+    to the configuration in their processing order. This means that a
+    configuration file named '00-default' has a lower priority than a file
+    named '01-my-overwrite'.
+    The global config file, either provided via '--config,-c' or per default in
+    %s, always has a lower priority than those inside the
+    '--config-dir,-d'.
+    Beside that, the provided command line parameters still have the highest
+    configuration priority.`, libconfig.CrioConfigPath),
```
Comment (TomSweeneyRedHat) [inline internal/pkg/criocli/criocli.go]: see prior comments.

File: docs/crio.8.md (hunk 6 above)
Comment (saschagrunert) [reply docs/crio.8.md]: Changed it to:
```
	Beside that, provided command line parameters still have a higher priority
	than any configuration file.
```
I hope this makes more sense.

**Pull request merged**